man/man1/now_playing.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "NOW_PLAYING" "1" "April 24, 2022" "now_playing 2.0.1" "User Manual"
.hy
.SH NAME
.PP
now_playing - Get track, artist, and album now playing in specified Roon
zone
.SH SYNOPSIS
.PP
\f[B]now_playing\f[R] [ \f[B]-a\f[R] ] [ \f[B]-h\f[R] ] [ \f[B]-p\f[R] ]
[ \f[B]-P\f[R] ] [ \f[B]-z\f[R] ZONE ]
.SH DESCRIPTION
.PP
Retrieves now playing information on specified zone or all zones if no
zone is specified.
Only actively playing zones are displayed if \f[B]-p\f[R] is provided,
only paused zones if \f[B]-P\f[R] is provided.
If the \f[B]-a\f[R] flag is provided then all zone states are displayed
regardless of their state.
.SH COMMAND LINE OPTIONS
.TP
\f[B]-a\f[R]
Display all zone states
.TP
\f[B]-h\f[R]
Display a usage message and exit
.TP
\f[B]-p\f[R]
Display only zones in the playing state
.TP
\f[B]-P\f[R]
Display only zones in the paused state
.TP
\f[B]-z ZONE\f[R]
If a \f[I]ZONE\f[R] is specified then retrieve now playing information
for that zone only.
If no \f[I]ZONE\f[R] is provided on the command line then retrieve now
playing information for all zones.
\[aq]-z default\[aq] indicates the default zone, \[aq]-z last\[aq] the
last zone used
.SH EXAMPLES
.TP
\f[B]now_playing\f[R]
Displays now playing info for all zones with active playing state
.TP
\f[B]now_playing -z Kitchen\f[R]
Displays now playing info for the Roon zone named \[dq]Kitchen\[dq]
.TP
\f[B]now_playing -a\f[R]
Displays now playing info for all zones regardless of state
.TP
\f[B]now_playing -p\f[R]
Displays now playing info for all zones in the playing state
.SH AUTHORS
.PP
Written by Ronald Record <github@ronrecord.com>
.SH LICENSING
.PP
NOW_PLAYING is distributed under an Open Source license.
See the file LICENSE in the NOW_PLAYING source distribution for
information on terms & conditions for accessing and otherwise using
NOW_PLAYING and for a DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<https://github.com/doctorfree/RoonCommandLine/issues>
.SH SEE ALSO
.PP
\f[B]get_core_ip\f[R](1), \f[B]get_zone_info\f[R](1),
\f[B]get_zones\f[R](1), \f[B]list_albums\f[R](1),
\f[B]list_artist_albums\f[R](1), \f[B]list_artists\f[R](1),
\f[B]list_composers\f[R](1), \f[B]list_genre_albums\f[R](1),
\f[B]list_genre_artists\f[R](1), \f[B]list_genres\f[R](1),
\f[B]list_playlists\f[R](1), \f[B]list_radio\f[R](1),
\f[B]list_tags\f[R](1), \f[B]list_zones\f[R](1),
\f[B]play_album\f[R](1), \f[B]play_artist\f[R](1),
\f[B]play_artist_album\f[R](1), \f[B]play_artist_track\f[R](1),
\f[B]play_composer\f[R](1), \f[B]play_genre\f[R](1),
\f[B]play_genre_album\f[R](1), \f[B]play_genre_artist\f[R](1),
\f[B]play_playlist\f[R](1), \f[B]play_radio\f[R](1),
\f[B]play_tag\f[R](1), \f[B]roon\f[R](1), \f[B]set_zone\f[R](1),
\f[B]set_zone_group\f[R](1), \f[B]zone_command\f[R](1)
.PP
Full documentation and sources at:
<https://github.com/doctorfree/RoonCommandLine>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

May 30, 2024 :
RoonCommandLine version 2.1.3 release 1

    This release provides support for `-z default` and `-z last` to specify default zone and last zone used
        * Add usage to `get_zone_info`
        * Updated man pages
        * Add `-P` flag to `now_playing` to list paused zones
        * Upgrade checks to prevent malformed `roon_api.ini`

May 27, 2024 :
RoonCommandLine version 2.1.2 release 4

    This release provides improved info and formatting for now playing and media search
        * `now_playing` info includes state
        * `now_playing` output formatted with `rich` if available
        * Fixes for upgrades on systems with BSD sed (e.g. macOS)
        * Media search and selection enhancements

May 23, 2024 :
RoonCommandLine version 2.1.2 release 3

    This release provides support for signal handling and fuzzy track search in menus
        * Add signal handler to Python scripts
        * Add tracks listing, track play, and album track play commands
        * Add Examples and Usage to main menu
        * Improve info when selecting and listing in menus
        * Install `jq` in `macInstall`

May 20, 2024 :
RoonCommandLine version 2.1.2 release 2

    This release provides support for installation and use of `roon-tui`
        * Install `roon-tui` with `roon -I`
        * Add `roon-tui` wrapper `roontui`
        * Add shortcuts to menus
        * Add `list_album_tracks` command
        * Use `fzf` to select media and list media

May 16, 2024 :
RoonCommandLine version 2.1.1 release 1

    This release provides support for shuffled artist/genre/playlist playback
        * Shuffled playback is accomplished using the play_media `Shuffle` action
        * All play commands now accept a `-z ZONE` argument to specify playback zone

May 09, 2024 :
RoonCommandLine version 2.1.0 release 6

    This release provides a more robust Python virtual environment installation of pip:
        * If ensurepip Python module not available then download and run get-pip.py
        * Requires curl or wget

May 08, 2024 :
RoonCommandLine version 2.1.0 release 5

    This release ensures that pip is installed in the Python virtual environment:
        * Activate the Python virtual environment in all install scripts
        * Use the built-in ensurepip Python module to ensure pip is installed
        * Fix timing after first discovery to avoid requiring dual authorization

Feb 20, 2024 :
RoonCommandLine version 2.1.0 release 4

    This release adds the capability to transfer the current queue to another zone:
        * Add `transfer_zone` command and Python backend
        * Stop core discovery after core discovered

Feb 14, 2024 :
RoonCommandLine version 2.1.0 release 3

    This release provides support for Ubuntu 23.04, Debian 12, and later:
        * Use a Python virtual environment
        * Additional checks for valid saved roon_api.ini
        * Add support for `q` to quit
        * Migrate to Github
        * Add lolcat colored figlet text transparent PNG header

May 16, 2023 :
RoonCommandLine version 2.0.8 release 5

    This is a bug fix release:
        * PEP 8 style guide conformance
        * Improved interactive menu system
        * Fix shellcheck warnings
        * Fix bug in call to get_zones in Get/Set Zone

May 2, 2023 :
RoonCommandLine version 2.0.8 release 4

    This is a bug fix release:
        * Only restore saved roon_api.ini if in new format

April 29, 2023 :
RoonCommandLine version 2.0.8 release 3

    This release adds support for:
        * Check for python3 in roon set_default
        * Support for Homebrew installed Python etc
        * Don't error out in postinstall
        * Postinstall fails when there is a directory that is not a valid user
        * Check if run as root user in Install
        * Always upgrade roonapi during install
        * Mac installation fixes
        * Restore ensurepip in Mac install
        * Fix permissions errors during Mac install
        * Remove symlinks in /usr/local/bin in Mac uninstall
        * Update install size in Debian packaging
        * Redirect stderr when invoking tput to set BOLD and NORMAL text
        * In Linux and Mac manual install scripts, upgrade roonapi if installed
        * Update set_volume man page, changelog, and release notes
        * Support percentage volume change using new roonapi methods
        * Handle negative volume levels in roon_fade

January 19, 2023 :
RoonCommandLine version 2.0.8 release 2 includes Mac installation fixes

December 31, 2022 :
RoonCommandLine version 2.0.8 release 1 provides API compatibility and feature updates

    This release adds support for:
        * Compatibility with Python Roon API 0.1.3
        * Support for percentage volume change/setting
        * Add immediate mode to roon_fade with `now` argument
        * Implement non-interactive immediate fading
        * Add interactive menu system (`roon_fade -i`)
        * Set man permissions correctly in manual install scripts

December 11, 2022 :
RoonCommandLine version 2.0.7 release 1 is a minor feature and bugfix update

    This release adds support for:
        * Save authentication token right after we get a RoonApi handle (bug fix)
        * Add options to `roon` command for controlling fading
        * Add fade feature (`roon_fade` user command and `roon_faded` fade daemon)
        * Man pages for fading and new commands
        * Fixes for `now_playing`, `get_zone_volume`, and `get_zone_attributes`
        * Add `get_zone_remaining` to get remaining time for now playing in zone
        * Add `get_zone_volume` and `get_zone_attributes.py`

November 15, 2022 :
RoonCommandLine version 2.0.6 release 2 is a minor feature update

    This release adds support for:
        * Add list_playlist_tracks to list a playlist's tracks
        * Run roonapi commands in postinstall as user if provided

api/now_playing.py
import argparse
import configparser
import json
from os import path
import signal
import sys
from roonapi import RoonApi


def signal_handler(sig, frame):
    sys.exit(0)


signal.signal(signal.SIGINT, signal_handler)

config = configparser.ConfigParser()
config.read("/usr/local/Roon/etc/roon_api.ini")

# Set to IP address of your Roon Core
server = config["DEFAULT"]["RoonCoreIP"]
# Set to Port of your Roon Core
port = config["DEFAULT"]["RoonCorePort"]
# Name of the file that holds a Roon API token
tokenfile = config["DEFAULT"]["TokenFileName"]

version = config["DEFAULT"]["RoonCommandLineVersion"]
release = config["DEFAULT"]["RoonCommandLineRelease"]
fullver = version + "-" + release

parser = argparse.ArgumentParser()
parser.add_argument("-z", "--zone", help="zone selection")
parser.add_argument(
    "-P",
    "--paused",
    default=False,
    action="store_true",
    help="display only zones in paused state",
)
parser.add_argument(
    "-p",
    "--playing",
    default=False,
    action="store_true",
    help="display only zones in playing state",
)
parser.add_argument(
    "-a",
    "--all",
    default=False,
    action="store_true",
    help="display all zones regardless of state",
)
args = parser.parse_args()

if args.zone:
    target_zone = args.zone
    if target_zone == "__all__":
        target_zone = None
else:
    target_zone = None
if args.paused:
    filterpaused = 1
else:
    filterpaused = None
if args.playing:
    filterplaying = 1
else:
    filterplaying = None
if args.all:
    filterplaying = None
    filterpaused = None

appinfo = {
    "extension_id": "roon_command_line",
    "display_name": "Python library for Roon",
    "display_version": fullver,
    "publisher": "RoonCommandLine",
    "email": "roon@ronrecord.com",
    "website": "https://github.com/doctorfree/RoonCommandLine",
}

# Can be None if you don't yet have a token
if path.exists(tokenfile):
    token = open(tokenfile).read()
else:
    token = "None"

with RoonApi(appinfo, token, server, port) as roonapi:

    # save the token for next time
    with open(tokenfile, "w") as f:
        f.write(str(roonapi.token))

    outputs = roonapi.outputs

    output_id = None
    if target_zone is None:
        for zone in roonapi.zones.values():
            state = "Unknown"
            if zone["state"] is not None:
                state = zone["state"]
            if filterplaying is not None and state != "playing":
                continue
            else:
                if filterpaused is not None and state != "paused":
                    continue
                else:
                    if "display_name" in zone:
                        zone_name = zone["display_name"]
                    else:
                        zone_name = "Unknown"
                    if "now_playing" in zone and zone["now_playing"] is not None:
                        print("\nNow playing in zone: %s" % zone_name)
                        track = json.dumps(
                            zone["now_playing"]["three_line"]["line1"],
                            ensure_ascii=False,
                        ).encode("utf8")
                        print("\tTrack:\t %s" % track.decode())
                        artist = json.dumps(
                            zone["now_playing"]["three_line"]["line2"],
                            ensure_ascii=False,
                        ).encode("utf8")
                        print("\tArtist:\t %s" % artist.decode())
                        album = json.dumps(
                            zone["now_playing"]["three_line"]["line3"],
                            ensure_ascii=False,
                        ).encode("utf8")
                        print("\tAlbum:\t %s" % album.decode())
                        print("\tState:\t %s" % state)
    else:
        for k, v in outputs.items():
            if target_zone in v["display_name"]:
                state = "Unknown"
                zone_name = v["display_name"]
                output_id = k
                zone = roonapi.zone_by_output_id(output_id)
                if zone is not None:
                    if zone["state"] is not None:
                        state = zone["state"]
                    if filterplaying is not None and state != "playing":
                        continue
                    else:
                        if filterpaused is not None and state != "paused":
                            continue
                        else:
                            if zone["now_playing"] is not None:
                                print("\nNow playing in zone: %s" % zone_name)
                                track = json.dumps(
                                    zone["now_playing"]["three_line"]["line1"],
                                    ensure_ascii=False,
                                ).encode("utf8")
                                print("\tTrack:\t %s" % track.decode())
                                artist = json.dumps(
                                    zone["now_playing"]["three_line"]["line2"],
                                    ensure_ascii=False,
                                ).encode("utf8")
                                print("\tArtist:\t %s" % artist.decode())
                                album = json.dumps(
                                    zone["now_playing"]["three_line"]["line3"],
                                    ensure_ascii=False,
                                ).encode("utf8")
                                print("\tAlbum:\t %s" % album.decode())
                                print("\tState:\t %s" % state)
